src/main/java/com/google/api/codegen/discovery/transformer/java/JavaSampleTypeNameConverter.java

```diff
@@ -0,0 +1,205 @@
+import com.google.api.codegen.util.TypedValue;
+import com.google.api.codegen.util.java.JavaTypeTable;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.Field;
+
+/**
+ * Maps SampleConfig and TypeInfo instances to Java specific TypeName instances.
+ */
+class JavaSampleTypeNameConverter implements SampleTypeNameConverter {
+
+  /**
+   * A map from primitive types in proto to Java counterparts.
+   */
+  private static final ImmutableMap<Field.Kind, String> PRIMIVITVE_TYPE_MAP =
+      ImmutableMap.<Field.Kind, String>builder()
+          .put(Field.Kind.TYPE_BOOL, "boolean")
+          .put(Field.Kind.TYPE_INT32, "int")
+          .put(Field.Kind.TYPE_INT64, "long")
+          .put(Field.Kind.TYPE_UINT32, "int")
+          .put(Field.Kind.TYPE_UINT64, "long")
+          .put(Field.Kind.TYPE_FLOAT, "float")
+          .put(Field.Kind.TYPE_DOUBLE, "double")
+          .put(Field.Kind.TYPE_STRING, "java.lang.String")
+          .put(Field.Kind.TYPE_ENUM, "java.lang.String")
+          .build();
+
+  /**
+   * A map from primitive types in proto to zero value in Java.
+   */
+  private static final ImmutableMap<Field.Kind, String> PRIMITIVE_ZERO_VALUE =
+      ImmutableMap.<Field.Kind, String>builder()
+          .put(Field.Kind.TYPE_BOOL, "false")
+          .put(Field.Kind.TYPE_INT32, "0")
+          .put(Field.Kind.TYPE_INT64, "0L")
+          .put(Field.Kind.TYPE_UINT32, "0")
+          .put(Field.Kind.TYPE_UINT64, "0L")
+          .put(Field.Kind.TYPE_FLOAT, "0.0F")
+          .put(Field.Kind.TYPE_DOUBLE, "0.0")
+          .put(Field.Kind.TYPE_STRING, "\"\"")
+          .put(Field.Kind.TYPE_ENUM, "\"\"")
+          .build();
+
+  private TypeNameConverter typeNameConverter;
+
+  private final String apiNameVersion;
+  /**
+   * Upper-case conversion of apiTypeName from SampleConfig.
+   */
+  private final String apiTypeName;
+  /**
+   * Upper-case conversion of methodNameComponents from SampleConfig.
+   */
+  private final List<String> methodNameComponents;
+
+  public JavaSampleTypeNameConverter(SampleConfig sampleConfig, String methodName) {
+    // We store these values to facilitate import path generation.
+    this.typeNameConverter = new JavaTypeTable();
+    String apiTypeNameOverride = sampleConfig.apiTypeNameOverride();
+    if (!Strings.isNullOrEmpty(apiTypeNameOverride)) {
+      this.apiTypeName = typeNameConverter.getTypeName(apiTypeNameOverride).getNickname();
+    } else {
+      this.apiTypeName = Name.lowerCamel(sampleConfig.apiName()).toUpperCamel();
+    }
+    this.apiNameVersion = sampleConfig.apiNameVersion();
+    List<String> copy = new ArrayList<>(sampleConfig.methods().get(methodName).nameComponents());
+    // Convert the method name components to upper camel case.
+    for (int i = 0; i < copy.size(); i++) {
+      copy.set(i, Name.lowerCamel(copy.get(i)).toUpperCamel());
+    }
+    this.methodNameComponents = copy;
```

Comment:
It's a bit tricky... these values are used in generating all message typenames and the api typename.

Since the api type and request type in java start with
`com.google.api.services.{apiname}.{apiversion}.{apitypename}[.{components}]`

And all other messages start with
`com.google.api.services.{apiname}.{apiversion}.model.{messagetypename}`

I thought it made sense to store all those values in the constructor, since they're important to generating fully qualified paths.

I think you're talking about moving the naming logic (lower -> upper) into samplenamer, which makes sense. I'll do that.